filters/auth/oidc.go

```diff
@@ -860,6 +922,120 @@ func (f *tokenOidcFilter) getTokenWithExchange(state *OauthState, ctx filters.Fi
 	return oauth2Token, err
 }
 
+// handleDistributedClaims handles if user has a distributed / overage token.
+// https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens#groups-overage-claim
+// In Azure, if you are indirectly member of more than 200 groups, they will
+// send _claim_names and _claim_sources instead of the groups, per OIDC Core 1.0, section 5.6.2:
+// https://openid.net/specs/openid-connect-core-1_0.html#AggregatedDistributedClaims
+// Example:
+//
+// {
+// 	 "_claim_names": {
+// 	   "groups": "src1"
+// 	 },
+// 	 "_claim_sources": {
+// 	   "src1": {
+// 	     "endpoint": "https://graph.windows.net/.../getMemberObjects"
+// 	   }
+//   }
+// }
+//
+func (f *tokenOidcFilter) handleDistributedClaims(ctx context.Context, oidcIDToken string, oauth2Token *oauth2.Token,
+	claimsMap map[string]interface{}) error {
+	idToken, err := f.verifier.Verify(ctx, oidcIDToken)
+	if err != nil {
+		return fmt.Errorf("failed to verify id token: %w", err)
+	}
+
+	// https://github.com/coreos/go-oidc/issues/171#issuecomment-1044286153
+	var distClaims distributedClaims
+	err = idToken.Claims(&distClaims)
+	if err != nil || distClaims.ClaimNames == nil || distClaims.ClaimSources == nil {
+		log.Debugf("No distributed claims found: %v", err)
+		return nil
+	}
+
+	claimMapSources := make(map[string]claimSource, len(distClaims.ClaimNames))
+	for claim, pointer := range distClaims.ClaimNames {
+		claimMapSources[claim] = distClaims.ClaimSources[pointer]
+	}
+
+	for claim, source := range claimMapSources {
+		uri, err := url.Parse(source.Endpoint)
```

Comment:
Do we need this extra `claimMapSources`+loop? It also assumes `ClaimSources` always has a `pointer` key without checking. Maybe it should be like this:
```go
for claim, sourceName := range distClaims.ClaimNames {
  source, ok := distClaims.ClaimSources[sourceName]
  if !ok {
      return fmt.Errorf("invalid distributed claims: missing claim source for %s", claim)
  }
  uri, err := url.Parse(source.Endpoint)
  ...
}
```

Reply:
SGTM